docs/specs/user-model.md

```diff
@@ -312,3 +294,164 @@ Once used, a recovery code is invalidated.
+
+## Claims Mapping
+
+A user has their own claims mapping. The claims mapping of a user can be updated via Admin API.
+
+> TODO: We may allow the user change some claims mapping, for example, we may allow the user to change which identity to map `email`.
+> However, it is unlikely that the developer would expect the user to change `custom_attributes` mapping.
+
+### Claims Mapping Example
+
+The following example illustrates the full capability of claims mapping.
+
+Given the following claims mapping
+
+```yaml
+mapping:
+- kind: "identity"
+  identity_id: "identity-a"
+  name_pointer: "#/email"
+  value_pointer: "#/email"
+- kind: "custom_attributes"
+  name_pointer: "#/zoneinfo"
+  value_pointer: "#/profile/preferred_timezone"
+- kind: "custom_attributes"
+  name_pointer: "#/picture"
+  value_pointer: "#/profile/profile_image_url"
+- kind: "custom_attributes"
+  name_pointer: "#/app:rbac"
+  value_pointer: "#/rbac"
+```
+
+And the claims of `identity-a`
+
+```json
+{
+  "email": "user@example.com"
+}
+```
+
+And the custom attributes
+
+```json
+{
+  "profile": {
+    "preferred_timezone": "Asia/Hong_Kong",
+    "profile_image_url": "https://cdn.example.com/u/user-a.jpg"
+  },
+  "rbac": [
+    "product:list",
+    "product:get",
+    "product:delete"
+  ]
+}
+```
+
+The computed claims of the user is
+
+```json
+{
+  "email": "user@example.com",
+  "zoneinfo": "Asia/Hong_Kong",
+  "picture": "https://cdn.example.com/u/user-a.jpg",
+  "app:rbac": [
+    "product:list",
+    "product:get",
+    "product:delete"
+  ]
+}
+```
+
+### Automatic addition of claims mapping
+
+When a new identity is being added, for each Standard Claims of the identity, if that claim is not mapped, then a new mapping is created from the claim.
+
+For example, if a new user signs up with Google, they will have the following claim mappings
+
+```
+mapping:
+- kind: "identity"
+  identity_id: "google"
+  name_pointer: "#/email"
+  value_pointer: "#/email"
+```
+
+So the user will have the claim `email` immediately available in their claims.
+
+### Dangling reference in claims mapping
+
+If a deletion of identity will result in dangling reference in the claims mapping, the deletion is disallowed.
+
+If a change in the custom attributes will result in dangling reference, the deletion is allowed, the mapping does not produce error and is no-op.
+
+### Claims mapping template
+
+The developer can define a claims mapping template to be copied to new user in the configuration.
+
+For example, in the configuration
+
+```yaml
+claims:
+  mapping:
+    template:
+    - kind: "custom_attributes"
+      name_pointer: "#/email"
+      value_pointer: "#/profile/email"
+    - kind: "custom_attributes"
+      name_pointer: "#/zoneinfo"
+      value_pointer: "#/profile/preferred_timezone"
+    - kind: "custom_attributes"
+      name_pointer: "#/picture"
+      value_pointer: "#/profile/profile_image_url"
+```
+
+The template is copied to the claims mapping of new user.
```

Comment:
If the template is copied, does that meant when the template is updated, it need to run migration for each users?

Reply:
Or easier to just make it as template for every users... (or how useful is it for each user have its own claims map?)

Reply:
> If the template is copied, does that meant when the template is updated, it need to run migration for each users?

Yes :(

Or we can improve the design by replacing the copy behavior with a runtime merge. So a change in the configuration could affect the claims of ALL users which could be considered as surprising.

---

> Or easier to just make it as template for every users... (or how useful is it for each user have its own claims map?)

The primary reason of user owning their claims mapping is that we support `kind: identity` claim mapping, which is a reference to one of the identity the user owns. This cannot be expressed in the claim mapping template in the configuration.

Reply:
What about if we always have template claims mapping applied to all users, and each users have their own claims mapping which "overwrite" the default if conflicted?

I think copy / runtime merge might be too complicated? I think the change in the configuration that does not affect the claims of ALL users would be surprising in the real world, because most use cases would expect the whole auth system (or at least for each client-id) have only 1 claims mapping...